x-pack/metricbeat/module/aws/s3_daily_storage/s3_daily_storage.go

```diff
@@ -0,0 +1,220 @@
+		if err != nil {
+			err = errors.Wrap(err, "GetMetricDataResults failed, skipping region "+regionName)
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		// Create Cloudwatch Events for s3_daily_storage
+		bucketNames := getBucketNames(listMetricsOutputs)
+		for _, bucketName := range bucketNames {
+			event, err := createCloudWatchEvents(metricDataOutputs, regionName, bucketName)
+			if err != nil {
+				m.logger.Error(err.Error())
+				event.Error = err
+				report.Event(event)
+				continue
+			}
+			report.Event(event)
+		}
+	}
+}
+
+func getBucketNames(listMetricsOutputs []cloudwatch.Metric) (bucketNames []string) {
+	for _, output := range listMetricsOutputs {
+		for _, dim := range output.Dimensions {
+			if *dim.Name == "BucketName" {
+				if aws.StringInSlice(*dim.Value, bucketNames) {
+					continue
+				}
+				bucketNames = append(bucketNames, *dim.Value)
+			}
+		}
+	}
+	return
+}
+
+func constructMetricQueries(listMetricsOutputs []cloudwatch.Metric, periodInSec int) []cloudwatch.MetricDataQuery {
+	metricDataQueries := []cloudwatch.MetricDataQuery{}
+	metricDataQueryEmpty := cloudwatch.MetricDataQuery{}
+	metricNames := []string{"NumberOfObjects", "BucketSizeBytes"}
+	for i, listMetric := range listMetricsOutputs {
+		if !aws.StringInSlice(*listMetric.MetricName, metricNames) {
+			continue
+		}
+
+		metricDataQuery := createMetricDataQuery(listMetric, periodInSec, i)
+		if metricDataQuery == metricDataQueryEmpty {
+			continue
+		}
+		metricDataQueries = append(metricDataQueries, metricDataQuery)
+	}
+	return metricDataQueries
+}
+
+func createMetricDataQuery(metric cloudwatch.Metric, periodInSec int, index int) (metricDataQuery cloudwatch.MetricDataQuery) {
+	statistic := "Average"
+	period := int64(periodInSec)
+	id := "s3d" + strconv.Itoa(index)
+	metricDims := metric.Dimensions
+	bucketName := ""
+	storageType := ""
+	for _, dim := range metricDims {
+		if *dim.Name == "BucketName" {
+			bucketName = *dim.Value
+		} else if *dim.Name == "StorageType" {
+			storageType = *dim.Value
+		}
+	}
+	metricName := *metric.MetricName
+	label := bucketName + " " + storageType + " " + metricName
+
+	metricDataQuery = cloudwatch.MetricDataQuery{
+		Id: &id,
+		MetricStat: &cloudwatch.MetricStat{
+			Period: &period,
+			Stat:   &statistic,
+			Metric: &metric,
+		},
+		Label: &label,
+	}
+	return
+}
+
+func createCloudWatchEvents(outputs []cloudwatch.MetricDataResult, regionName string, bucketName string) (event mb.Event, err error) {
+	event.Service = metricsetName
+	event.RootFields = common.MapStr{}
+	// Cloud fields in ECS
+	event.RootFields.Put("service.name", metricsetName)
+	event.RootFields.Put("cloud.region", regionName)
+
+	// AWS s3_daily_storage metrics
+	mapOfMetricSetFieldResults := make(map[string]interface{})
+	for _, output := range outputs {
+		labels := strings.Split(*output.Label, " ")
+		if labels[0] == bucketName {
```

Comment:
I'd check for empty slice here before accesing first element, just in case AWS changes something or some network error mess things up

Reply:
done 😄